presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java

```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
             throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Concatenated string is too large");
         }
     }
+
+    private static class ConcatSignatureBinder
+            extends SignatureBinder
+    {
+        public ConcatSignatureBinder(TypeManager typeManager, Signature signature, boolean allowCoercion)
+        {
+            super(typeManager, signature, allowCoercion);
+        }
+
+        @Override
+        public Optional<Signature> bind(List<? extends Type> actualArgumentTypes)
+        {
+            ImmutableList.Builder<TypeSignature> boundTypesSignatures = ImmutableList.builder();
+
+            long lengthAccumulator = 0;
+            for (Type type : actualArgumentTypes) {
+                if (!isTypeValid(type)) {
+                    return Optional.empty();
+                }
+
+                int paramLength = varcharLength(type);
```

Comment:
You do not take coercions into account in `varcharLength`. And `isTypeValid` allows ending up here with type being anything coercible to `varchar`. It seems not right to me.